node/src/actors/chain_manager/handlers.rs

```diff
@@ -494,19 +494,25 @@ impl Handler<AddBlocks> for ChainManager {
 
                                     // We must construct the second superblock in order to be able
                                     // to validate the votes for this superblock later
-                                    log::debug!("Will construct the second superblock during synchronization. Superblock index: {} Epoch {}", sync_target.superblock.checkpoint + 1, candidate_superblock_epoch);
+                                    log::debug!("Will construct the second superblock during synchronization. Superblock index: {} Epoch {}", sync_target.superblock.checkpoint + 1, candidate_superblock_checkpoint * superblock_period);
 
-                                    actix::fut::ok(candidate_superblock_epoch)
+                                    actix::fut::ok(candidate_superblock_checkpoint)
                                 }
                             })
-                            .and_then(move |candidate_superblock_epoch, act, ctx| {
-                                act.build_and_vote_candidate_superblock(ctx, candidate_superblock_epoch).map(move |_, _, _| candidate_superblock_epoch)
+                            .and_then(move |candidate_superblock_checkpoint, act, ctx| {
+                                if let Some(candidate_superblock_epoch) = act.superblock_candidate_is_needed(candidate_superblock_checkpoint, superblock_period) {
+                                    actix::fut::Either::A(act.build_and_vote_candidate_superblock(ctx, candidate_superblock_epoch).map(move |_, _, _| Some(candidate_superblock_epoch)))
+                                }
+                                else{
+                                    actix::fut::Either::B(actix::fut::ok(None))
+                                }
                             })
                             .and_then(move |candidate_superblock_epoch, act, ctx| {
-                                // Store chain state at candidate superblock, but do not persist it yet
-                                let superblock_index = candidate_superblock_epoch / superblock_period;
-                                act.move_chain_state_forward(superblock_index);
-
+                                if let Some(candidate_superblock_epoch) = candidate_superblock_epoch {
```

Comment:
Now that I see it, this if can be combined with the previous one, and we don't need to return `candidate_superblock_epoch` anymore.